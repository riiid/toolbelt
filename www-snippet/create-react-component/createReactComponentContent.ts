import { CreateReactComponentOption } from "./CreateReactComponentOption.ts";

export const createReactComponentContent = (
  name: string,
  option: CreateReactComponentOption,
) => {
  if (!option.forwardRef) {
    return `import React, { memo, PropsWithChildren } from "react";

export interface ${name}Props {}

${
      option.onlyNamedExport
        ? "export const"
        : "const"
    } ${name} = memo((({}: PropsWithChildren<${name}Props>) => {
  return <>${name}</>;
}));

${option.onlyNamedExport === false ? `export default ${name};` : ""}
`;
  }

  return `import React, { forwardRef, memo, HTMLAttributes } from "react";

type RefElement = HTMLDivElement;
type RefElementProps = HTMLAttributes<HTMLDivElement>;

export interface ${name}Props extends RefElementProps {}

${
    option.onlyNamedExport
      ? "export const"
      : "const"
  } ${name} = memo(forwardRef<RefElement, ${name}Props>((props, ref) => {
  return <div ref={ref} {...props}>Lorem Ipsum</div>;
}));

${option.onlyNamedExport === false ? `export default ${name};` : ""}

${name}.displayName = "${name}";
`;
};
