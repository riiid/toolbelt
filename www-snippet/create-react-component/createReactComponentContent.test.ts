import {
  assertSnapshot,
} from "https://deno.land/std@0.138.0/testing/snapshot.ts";
import { createReactComponentContent } from "./createReactComponentContent.ts";

Deno.test(
  `should return correct react component.`,
  async (t) => {
    await assertSnapshot(
      t,
      createReactComponentContent("LoremIpsum", {
        forwardRef: false,
        onlyNamedExport: false,
      }),
    );

    await assertSnapshot(
      t,
      createReactComponentContent("LoremIpsum", {
        forwardRef: true,
        onlyNamedExport: false,
      }),
    );
    await assertSnapshot(
      t,
      createReactComponentContent("LoremIpsum", {
        forwardRef: false,
        onlyNamedExport: true,
      }),
    );

    await assertSnapshot(
      t,
      createReactComponentContent("LoremIpsum", {
        forwardRef: true,
        onlyNamedExport: true,
      }),
    );
  },
);
