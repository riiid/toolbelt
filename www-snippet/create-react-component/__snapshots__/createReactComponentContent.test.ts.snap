export const snapshot = {};

snapshot[`should return correct react component. 1`] = `
'import React, { memo, PropsWithChildren } from "react";
export interface LoremIpsumProps {}

const LoremIpsum = (({}: PropsWithChildren<LoremIpsumProps>) => {
  return <>LoremIpsum</>;
});

export default memo(LoremIpsum);
'
`;

snapshot[`should return correct react component. 2`] = `
'import React, { forwardRef, memo } from "react";

type RefElement = HTMLDivElement;
type RefElementProps = React.HTMLAttributes<HTMLDivElement>;

export interface LoremIpsumProps extends RefElementProps {}

const LoremIpsum = forwardRef<RefElement, LoremIpsumProps>((props, ref) => {
  return <div ref={ref} {...props}>Lorem Ipsum</div>;
});

export default memo(LoremIpsum)

LoremIpsum.displayName = "LoremIpsum";
'
`;

snapshot[`should return correct react component. 3`] = `
'import React, { memo, PropsWithChildren } from "react";
export interface LoremIpsumProps {}

const LoremIpsum = (({}: PropsWithChildren<LoremIpsumProps>) => {
  return <>LoremIpsum</>;
});

export { LoremIpsum: memo(LoremIpsum) };
'
`;

snapshot[`should return correct react component. 4`] = `
'import React, { forwardRef, memo } from "react";

type RefElement = HTMLDivElement;
type RefElementProps = React.HTMLAttributes<HTMLDivElement>;

export interface LoremIpsumProps extends RefElementProps {}

const LoremIpsum = forwardRef<RefElement, LoremIpsumProps>((props, ref) => {
  return <div ref={ref} {...props}>Lorem Ipsum</div>;
});

export { LoremIpsum: memo(LoremIpsum) }

LoremIpsum.displayName = "LoremIpsum";
'
`;
